:seedling: Initial Hub API Tests

aufi opened the pull request:
First part of Hub REST API test suite.

Steps
- the simplest CRUD for all resource types
- simple seeds tests
- connecting resources with refs (update tests above)
- clarify non-trivial fixtures definition and tear up&down
- basic integration scenarios (addons&analysis)

Related to https://github.com/konveyor/tackle2-hub/issues/262

jortel commented on test/api/application/create_test.go:
```diff
@@ -0,0 +1,70 @@
+	"github.com/konveyor/tackle2-hub/api"
+)
+
+func TestApplicationCreate(t *testing.T) {
```
Looks like every test function has an TestApplication prefix that is redundant in the `application` package.  Everything in the package is implied to be scoped to application.
Recommend:
For example: just `TestCreate()` instead of: `TestApplicationCreate()`.  (less stuttering please).

jortel commented on test/api/testclient/client.go:
```diff
@@ -0,0 +1,41 @@
+
+// Add with and without login/token creation
+func NewHubClient() (client *addon.Client, err error) {
+	client = addon.NewClient(os.Getenv("HUB_ENDPOINT"), "")
```
Let's consider using `HUB_BASE_URL` already defined in 'settings'.

jortel replied:
Also, why not just do all of this in package `init()` instead of a new client object and functions?
See what I suggested on the issue.

aufi replied:
Yes, will update it (the ```init()```method as suggested fits great for this purpose, I'm updating the (Test)Client package).

aufi commented on test/api/testclient/action.go:
```diff
@@ -0,0 +1,15 @@
+package testclient
```
Good point, will rename it (maybe with more updates on the Client package).

jortel commented on test/api/testclient/client.go:
```diff
@@ -0,0 +1,41 @@
+	return
+}
+
+type LoginObject struct {
```
Why not just use api.Login instead?

aufi commented on test/api/application/fixtures.go:
```diff
@@ -0,0 +1,70 @@
+
+//
+// Create an Application (and stop tests on failure).
+func Create(t *testing.T, application *api.Application) {
```
I did, partially because this becomes handy for the same resource testing (e.g.Create() before Update test and Delete() after the test). I'm likely not implementing other mathods than Create&Delete unless see a use-case.

aufi commented:
Related to: https://github.com/konveyor/tackle2-hub/issues/262

aufi commented on test/api/stakeholdergroup/api_test.go:
```diff
@@ -0,0 +1,86 @@
+	"github.com/konveyor/tackle2-hub/test/api/client"
+)
+
+func TestStakeholderGroupCRUD(t *testing.T) {
```
@jortel Would this way of naming variables etc in this test have better fit than more verbose (more different variable names depending on tested resource type) version started in application test?

jortel commented on test/api/client/client.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// Login performs a login request to the hub and returns a token
+func login(client *addon.Client, username, password string) (string, error) {
```
Seems like the login can be done in init().

aufi commented on test/api/application/list_test.go:
```diff
@@ -0,0 +1,37 @@
+	samples := Samples()
+	// Create.
+	for i := range samples {
+		Create(t, &samples[i]) // Need modify parent array instead of single record created by the loop (to keep created IDs).
```
Updated Create parameter to be the parent array's item to avoid ```range``` making the item copy (which is not a pointer anymore).

aufi commented on addon/adapter.go:
```diff
@@ -157,7 +157,7 @@ type Params map[string]interface{}
 //
 // inject values into path.
-func (p Params) inject(path string) (s string) {
+func (p Params) Inject(path string) (s string) {
```
To be able use the ```inject``` method in tests, it was needed make it public.

aufi commented:
A sample with RichClient (not fully working) to consider, but it goes slightly away from the API interaction.

https://github.com/aufi/tackle2-hub/commit/51789771ddd79002b4748a4173a254f1126c48c3

aufi commented:
@jortel @mansam @djzager @dymurray Hey! This is a draft for Hub API tests, we had few discussion with Jeff already, but I'd like ask you for review and your thoughts. Thanks!

We can setup a call to discuss it if needed.

dymurray commented on test/api/application/fixtures.go:
```diff
@@ -0,0 +1,52 @@
+
+//
+// Delete the Application (and stop tests on failure).
+func Delete(t *testing.T, r *api.Application) {
```
Recommend renaming this to `DeleteApplication`

dymurray reviewed (commented):
My suggestion:

`Samples()` function as it exists is okay, since you are returning an array of applications to be used for tests. However, it feels like we are *close* to table driven testing without actually doing it. I think a small tweak to each test can make it table-driven without too major of a change. You should be defining a set of tests at the beginning of each function, and if you want to pull from the samples to define an application for the test that is good.

See: https://github.com/konveyor/go-konveyor-tests/blob/main/developer/applications-inventory/applications_analyze_test.go#L14 how we define the test cases at the top in a table, and this can give you flexibility of defining test cases where you expect failures and things to go wrong rather than having to write a whole separate test function for it. What  do you think?

aufi commented:
> My suggestion:
> 
> `Samples()` function as it exists is okay, since you are returning an array of applications to be used for tests. However, it feels like we are _close_ to table driven testing without actually doing it. I think a small tweak to each test can make it table-driven without too major of a change. You should be defining a set of tests at the beginning of each function, and if you want to pull from the samples to define an application for the test that is good.
> 
> See: https://github.com/konveyor/go-konveyor-tests/blob/main/developer/applications-inventory/applications_analyze_test.go#L14 how we define the test cases at the top in a table, and this can give you flexibility of defining test cases where you expect failures and things to go wrong rather than having to write a whole separate test function for it. What do you think?

Thanks for the comment! Actually even this PR was started with the TestCases and shouldFail true/false structure (e.g. https://github.com/konveyor/tackle2-hub/pull/268/commits/82d9c32072876970983001927ae82c2cccea2cbb or https://github.com/konveyor/tackle2-hub/pull/268/commits/ae4d8d01b1ef9eb556ff57d24afa5aa0a06d4ab2). However it turned out that it doesn't have right fit for our (mostly API) tests use-case.

(WIP - going to add examples and objectives why did it in such way).

aufi commented:
Added richclient methods, README and sample API integration (slow) test doing an application analysis https://github.com/konveyor/tackle2-hub/pull/268/commits/d49f50de9e7b81ad18741735ccd302d79d754198#diff-fcd2bf711a00447192da2d46171f15d0bb6302397b523e0ba92ac9f61bbb8ff7R17-R88.

There are still few things open like update API tests with more table-driven instead of "samples-driven" CRUD tests (together with considering use Samples map instead of array). But as an initial version of Hub tests, I think this PR is ready for review.

jortel commented on addon/adapter.go:
```diff
@@ -5,10 +5,11 @@ Tackle hub/addon integration.
-	"os"
 )
 
 var (
```
We need to decide on import ordering.  Looks like your IDE may be organizing imports using _goimports_.  revert this change so we don't have the thrashing.

jortel commented on addon/adapter.go:
```diff
@@ -5,10 +5,11 @@ Tackle hub/addon integration.
 	"github.com/konveyor/controller/pkg/logging"
 	"github.com/konveyor/tackle2-hub/settings"
 	"golang.org/x/sys/unix"
-	"os"
```
We need to decide on import ordering.  Looks like your IDE may be organizing imports using _goimports_.  revert this change so we don't have the thrashing.

aufi replied:
Thanks, I missed this unwanted change, reverting.

On the formatting tool, happy to discuss it, for me personally, I prefer "separated" imports (from goimports), but e.g. not removing empty comment lines like goimports do by default (and that's what we use quite often in go files), but I can accept whatever we agree on.

jortel commented on test/api/client/client.go:
```diff
@@ -0,0 +1,53 @@
+// When using enabled auth, don't forget set AUTH_REQUIRED=1 to make Hub load Auth settings.
+func New() (client *addon.Client, err error) {
+	baseUrl := settings.Settings.Addon.Hub.URL
+	login := api.Login{User: settings.Settings.Auth.Keycloak.Admin.User, Password: settings.Settings.Auth.Keycloak.Admin.Pass}
```
When testing with auth, I don't think this is the user/pass we should be using but fine for now so long as it works.

jortel commented on test/api/client/client.go:
```diff
@@ -0,0 +1,53 @@
+	client = addon.NewClient(baseUrl, "")
+
+	// Disable HTTP requests retry for network-related errors to fail quickly.
+	client.Retry = 0
```
:+1:

jortel commented:
In each package, importing and creating local var/types really reduces code _cruft_.
Consider taking it a step further:
```
import (
	"github.com/konveyor/tackle2-hub/api"
	c "github.com/konveyor/tackle2-hub/test/api/client"
)

var (
	Client = c.Client
+	Path = c.Path
)

+ type Params = c.Params
```

Importing the client package as `c` introduces a single letter global which I found confusing at first. Why alias the import?

Perhaps:
```
import (
	"github.com/konveyor/tackle2-hub/api"
	"github.com/konveyor/tackle2-hub/test/api/client"
)

var (
	Client = client.Client
+	Path = client.Path
)

+ type Params = client.Params
```

jortel commented on test/README.md:
````diff
@@ -0,0 +1,69 @@
+```
+
+```
+$ export HUB_ENDPOINT=http://localhost:8080
````
HUB_BASE_URL

aufi commented:
> In each package, importing and creating local var/types really reduces code _cruft_. Consider taking it a step further:
> 
> ```
> import (
> 	"github.com/konveyor/tackle2-hub/api"
> 	c "github.com/konveyor/tackle2-hub/test/api/client"
> )
> 
> var (
> 	Client = c.Client
> +	Path = c.Path
> )
> 
> + type Params = c.Params
> ```
> 
> Importing the client package as `c` introduces a single letter global which I found confusing at first. Why alias the import?
> 
> Perhaps:
> 
> ```
> import (
> 	"github.com/konveyor/tackle2-hub/api"
> 	"github.com/konveyor/tackle2-hub/test/api/client"
> )
> 
> var (
> 	Client = client.Client
> +	Path = client.Path
> )
> 
> + type Params = client.Params
> ```

Thanks for this point! The aliased client import as "c" is originated in _test.go files where writing client.Must/client.Should as well as client.Path/client.Params looked too verbose to me and I wanted keep the client alias consistent across files.

The ideal solution from my point of view woudl be import test's client package methods as first-class citizens into a resource test package, but that either introduces code for importing each method (which looks to me a biolerplate) or is not possible in golang AFAIK.

I'm going to provide an alternative solution on this.

aufi commented on Makefile:
```diff
@@ -112,3 +114,18 @@ ifeq (,$(wildcard $(INSTALL_TACKLE_SH)))
 	$(INSTALL_TACKLE_SH);
+
+# Run Hub test suite.
+test: test-unit test-api test-integration
```
test -> test-unit;
test-all? - TBD

aufi commented on .github/workflows/main.yml:
```diff
@@ -37,6 +37,28 @@ jobs:
+        with:
+          go-version: '1.19'
+      - run: |
+          make run &
```
@djzager I'm not sure about this part, we might discuss it or add in follow-up PR.

aufi commented on test/api/application/fixtures.go:
```diff
@@ -0,0 +1,67 @@
+
+var (
+	// Setup Hub API client
+	Client = client.Client
```
@jortel Import aliases were removed, assertions moved to assert package. Removal of the Client variable setup leading to ```client.Client.Post(...)``` on many lines in tests did look too strange to me, so I left it as it was.

If we are strongly opinionated about removing Client variable when other non-client methods are called via their package names, I'd  consider some kind of rename client.Client to client.Api/API to be ```client.API.Post(...)``` that looks better to me.

aufi commented:
__test-e2e__ task seems to fail also in another jobs, there is likely a different issue in tackle install.

djzager commented on test/api/application/list_test.go:
```diff
@@ -0,0 +1,36 @@
+		assert.Must(t, Create(&samples[i])) // Need modify parent array instead of single record created by the loop (to keep created IDs).
+	}
+
+	// Try list.
+	got := []api.Application{}
+	err := Client.Get(api.ApplicationsRoot, &got)
+	if err != nil {
+		t.Errorf("List error: %v", err.Error())
+	}
```
This looks like the kind of thing that we don't need a specific `list_test.go` to verify. By this I mean, if you update create and delete test fixtures to list and verify the application is either in (or not in) the list, then we've verified `list` is working without churning through this scenario.

aufi replied:
I partially agree, but since this is an API test, I'd expect explicit test for each endpoint method.

aufi replied:
Will be updated merging create with get/list.

djzager commented on test/api/application/fixtures.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+//
+// Set of valid Application resources for tests and reuse.
+// Invalid application for negative tests are expected to be defined within the test methods, not here.
+func Samples() (samples []api.Application) {
+	samples = []api.Application{
+		{
+			Name:        "Pathfinder",
+			Description: "Tackle Pathfinder application.",
+			Repository: &api.Repository{
+				Kind:   "git",
+				URL:    "https://github.com/konveyor/tackle-pathfinder.git",
+				Branch: "1.2.0",
+			},
+		},
+		{
+			Name: "Minimal application",
+		},
+	}
+	return
+}
```
Maybe, instead of having `Samples()` like this. We create a public function that returns a single application. For example `MinimalApp()` or `PathfinderGit()` that can be reused.

It might not be meaningful now, but we could consider a helper function that accepts a slice of applications and just creates them. I would expect this would be helpful at the start of scenarios.

dymurray replied:
I prefer this as oppose to calling something like `Samples[0]` in places where you wish to use the Pathfinder sample.

aufi replied:
Used Map for fixtures for now. Will check if other options fit better in followup tests development.

aufi commented on .github/workflows/test-nightly.yml:
```diff
@@ -0,0 +1,32 @@
+
+on:
+  schedule:
+    - cron: '13 0,12 * * *' # Regulary every 12 hours
```
I'm OK removing nightly build from this PR, but I believe that schedule-based application-level tests should run within the upstream CI.

aufi replied:
Keep for now to see how it works

aufi commented on test/README.md:
````diff
@@ -0,0 +1,69 @@
+- Integration tests ```$ make test-integration```
+- WIP Export/import tests
+
+All tests can be executed with ```$ make test-all``` which will run all available tests.
````
The test-all target is still present in Makefile line 134. However its main purpose is not to be executed in github actions, but more likely locally if needed by someone (I use it locally).

aufi commented on test/api/application/bucket_test.go:
```diff
@@ -0,0 +1,20 @@
+	// Create the application.
+	assert.Must(t, Create(&application))
+
+	// Bucket test TODO// with Client.BucketGet, BucketPut etc.
```
Yes, each application get a bucket. The bucket provides more actions which its content that are expected to be tested within this test file.

aufi replied:
:+1: check the bucket object ref

aufi commented on test/api/application/delete_test.go:
```diff
@@ -0,0 +1,32 @@
+				t.Errorf(err.Error())
+			}
+
+			// Check the it was deleted.
+			err = Client.Get(rPath, &r)
+			if err == nil {
+				t.Errorf("Exits, but should be deleted: %v", r)
+			}
```
The responsibility of fixtures.go methods is to perform given action (and relay the error if returned), it doesn't know if the action should or should not raise the error.
The tests know what should be the result of calling fixtures methods (or direct API Get/Post/.. methods) and the test might not want rely on method's returned error, so it might do another cal to check the real result.
This allows both positive and negative tests and I believe, this _should not_ be part of fixtures.

djzager replied:
This comment makes me think we are in disagreement on the purpose of these fixtures. If you would like to update the `Create()` fixture to accept an `err` that represents the error that would be returned, that would be great. But the idea of the fixtures is to strengthen the verification for these core, always repeating actions, ie. everytime we create an application we make sure 1) we can GET it 2) we can LIST it 3) it's bucket is created as expected. This way, when some addon writes a test using these fixtures these verifications are happening for free.

It's important for fixtures to verify the thing that was supposed to happen happened.

aufi replied:
Probably the disagreement comes from different understanding what should be tested with different kinds of tests. The GET/LIST/BUCKET after Create makes sense to me for the API endpoint test, but e.g. the addon analyze test might not focus on the application creation action if it is bulletproof OK or not, the addon test should create the application (ideally by fixtures method) and if the Create method has not failed, use the application in its test.

aufi replied:
I'm not against adding a helper which will check the created/deleted/.. resource like application, I just don't see it as a mandatory part of the fixtures.go methods.

aufi commented on Makefile:
```diff
@@ -112,3 +114,21 @@ ifeq (,$(wildcard $(INSTALL_TACKLE_SH)))
+.PHONY: test test-api test-integration
+
+# Run unit tests.
+test:
+	go test -v ./auth/
```
Tests generaly might have special requirements to setup (like running Hub API, theoretically running keycloak or whole Konveyor) ,so I'm not sure about scanning directories for test files. To specify unit tests-only (=> no dependencies except the golang environment), I'd prefer add it to the Makefile test section for unit test execution.

aufi commented:
The PR was updated in last days, I'm about continue this work in folloup PRs, but would be great get this PR in (or request change if there is something not acceptable, thanks).

djzager commented on test/api/client/client.go:
```diff
@@ -0,0 +1,50 @@
+
+//
+// Merge path with params.
+func Path(base string, paramsMap map[string]interface{}) string {
```
Is there a technical reason you say this?

> test files should not import addon package

djzager commented on test/api/application/bucket_test.go:
```diff
@@ -0,0 +1,26 @@
+	application := Samples()["Minimal"]
+
+	// Create the application.
+	assert.Must(t, Create(&application))
```
So, here we use the helper function, but in the create_get_delete_test.go we use the raw client. Why?

aufi replied:
Focus in this test file was on the buket get, so it looked to me easier to use short command to create a application for the bucket test. (this is not part of the reuse test pkg, it is its usage within the API test, so it is expected that test author might decide what fits better for its use case, more desciption in part of https://github.com/konveyor/tackle2-hub/pull/268#issuecomment-1520072758)

djzager commented on test/api/client/client.go:
```diff
@@ -0,0 +1,53 @@
+
+var Client *addon.Client
+
+func init() {
+	var err error
+	Client, err = New()
+	if err != nil {
+		panic(fmt.Sprintf("Error: Cannot setup API client: %v.", err.Error()))
+	}
+}
```
I think we should leave the instantiation of the client to the consumer.

aufi replied:
Idea behind this is that since test packages (test/application, test/tag, etc.) provide methods for interaction with Hub API, it should be responsible for the client setup. And having this same code in each resource type test package would be a duplication, it is easier to set it up here directly within the client pkg.

djzager commented on test/api/client/client.go:
```diff
@@ -0,0 +1,53 @@
+// When using enabled auth, don't forget set AUTH_REQUIRED=1 to make Hub load Auth settings.
+func New() (client *addon.Client, err error) {
+	baseUrl := settings.Settings.Addon.Hub.URL
+	login := api.Login{User: settings.Settings.Auth.Keycloak.Admin.User, Password: settings.Settings.Auth.Keycloak.Admin.Pass}
```
Does this work if auth is disabled?

djzager replied:
I assume it must since I don't see any configuring of auth in the nightly you included.

aufi replied:
Yes, works with disabled auth.

djzager commented on test/api/task/api_test.go:
```diff
@@ -0,0 +1,88 @@
+
+	for name := range samples {
+		sample := samples[name]
+		assert.Must(t, Create(&sample))
```
It's really confusing that we have the addon client, the test client, and these helper methods all being used and it's not really clear when we use which.

aufi replied:
I'm going to update README mentioning usage of both addon and rich client methods.

djzager reviewed (changes requested):
Thank you for being patient with me. I'm beginning to wonder if this PR would have been easier to review if we had just started with CRUD on applications :sunglasses: 

After further review, there are a few things I struggle with more than  the rest:

1. What are the rules for when we are using the addon.Client, the internal Client, and the helper functions? Have we considered, especially given that we aren't going to do additional verification beyond did it return 200, just using the addon.Client directly? If we are going to keep the helper functions around, they should at least  grab the object ID off the response.
2. I'm still against the Samples() `map[string]api.Application`.
3. In the application test files, we have a lot of churn on creating and deleting applications. What's preventing us from creating a batch of applications, verifying for all of them their bucket was created and we can GET it, add some facts to an application, update another application, ..., profit.

aufi commented on test/api/application/fixtures.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/konveyor/tackle2-hub/test/api/client"
+)
+
+// Set of valid Application resources for tests and reuse.
+// Invalid application for negative tests are expected to be defined within the test methods, not here.
+func Samples() (samples map[string]api.Application) {
+	samples = map[string]api.Application{
+		"Minimal": {
+			Name: "Minimal application",
+		},
+		"PathfinderGit": {
+			Name:        "Pathfinder",
+			Description: "Tackle Pathfinder application.",
+			Repository: &api.Repository{
+				Kind:   "git",
+				URL:    "https://github.com/konveyor/tackle-pathfinder.git",
+				Branch: "1.2.0",
+			},
+		},
+
+	}
+	return
+}
```
I agree the ```application.MinimalApplication``` (or shorter  ```application.Minimal``` ) looks better. Downsides are
- there is no iterator over applications out of the box
- it is not possible modify applications before using it in the test, e.g. some time ago, test paralelization was mention, even it is not required now, the Samples() function have an ability modify e.g. Application names before returning it (like if app Name is not empty, append space+5 random chars to Name)

aufi commented on test/integration/applications-inventory/analysis/basic_test.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+//
+// Test application analysis
+// "Basic" means that there no other dependencies than the application itself (no need prepare credentials, proxy, etc)
+func TestBasicAnalysis(t *testing.T) {
```
IIRC, it was accepted was to begin with this test as part of Hub repo. Later or with more complicated integration/e2e test cases, do it in separate repo.

aufi commented on test/api/application/create_get_delete_test.go:
```diff
@@ -0,0 +1,109 @@
+	// Create on array of Applications calls subtest
+	for _, r := range samples {
+		t.Run(r.Name, func(t *testing.T) {
+			err := Client.Post(api.ApplicationsRoot, &r)
```
It looked to me to be more descriptive to use a command with path as parameter to be clearly visible what path is being tested.
 (this is not part of the reuse test pkg, it is its usage within the API test, so it is expected that test author might decide what fits better for its use case, more desciption in part of https://github.com/konveyor/tackle2-hub/pull/268#issuecomment-1520072758)

aufi commented:
@jortel, @djzager PR was updated based on yesterdays discussion - last 2 commits - details below. I'd like ask for you ack/comments to continue work this way.

A ```/binding``` package was added in following steps:
- copy-pasted client.go and error.go from addon
- modified addon adapter.go to binding richclient.go
- added  Application methods to binding/richclient based on test methods
- TODO: add other resources methods only ack-ed the way how application is done

```/test/api/application/...``` test files were updated to not use ```test/client``` (which will be removed before this PR goes in), but use the ```binding/richclient``` (+addon client from the richclient for few calls not yet supported by the richclient).

A question I'd like to confirm:
Existing addon package methods on resources structs (like Application) use ```func (h *Application) Get(id uint) (r *api.Application, err error)``` (https://github.com/konveyor/tackle2-hub/blob/main/addon/application.go#L16-L23) functions that initiate and _return_ the application.
However the addon's "raw" client returns values by modifying parameter passed as reference and the same do tests so far ```func (h *Application) Get(r *api.Application) (err error)``` (https://github.com/konveyor/tackle2-hub/pull/268/files#diff-d1be7262de63443ed32ecff32f7e38074f4c9569417ea354683fe4ec05704dd0R21-R24).
The "reference" way makes easy use of methods like ```assert.Should(t, <action>)```) that return only the error and code looks better for my point of view. A downside is little awkward way of Get method (https://github.com/konveyor/tackle2-hub/pull/268/files#diff-0eaa5b22ad02c862b6bd490ac8274d8bc5b1985949d6e3f193b9deabf1b33001R18), that needs to prepare struct where the response should be stored to and need to set its ID before doing the API call. And incompatibility with existing addon methods.

The question is: does the ```binding``` package need to implement addon's "helper" methods signature (initiate internally resources and return it) instead of using passed refs (like addon client and in tests now)?

jortel commented on binding/richclient.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	RichClient = newRichClient()
+}
```
I think we decided to not provide the `RichClient` global.
Please remove the `var` and the `init()`.
Please rename `Adapter` => `RichClient`.

aufi replied:
:+1: Updated (the init() was left for Settings load)

jortel commented on binding/richclient.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// newRichClient builds a new Addon RichClient object.
+func newRichClient() (richClient *Adapter) {
```
Let's rename _newRichClient()_ to just `New(url) (client *RichClient)`.
This will be the _main_ constructor for the package.

Add `RichClient.Login(user, password string) (err error)`

```
client := binding.New("http://localhost:8080/hub")
err := client.Login("elmer", "fudd")
...
```

aufi replied:
:+1: Updated.

jortel commented:
The binding should follow the same signature pattern as the _addon_ binding.
Example:
- Create(r *Application) (err error)
- Get(int id) (r *Application, err error)
- List() (list []api.Application, err error)
- Update(r *Application) (err error)
- Delete(int id) (err error)

In the future, the List() should also support _filter_.

aufi commented on binding/application.go:
```diff
@@ -0,0 +1,42 @@
+type Application struct {
+	// hub API client.
+	client *Client
+}
```
Ack, will update binding pkg&tests.

aufi replied:
Application methods in binding package were updated according to comment above.

aufi commented:
The PR was updated based on discussion from call we had last week with @dymurray @jortel @djzager @mansam and clarification from Jeff this week. The PR is up for review.

I'm about submitting follow-up PRs focusing on 
- covering most of resources in basic API test with the binding package update
- expansion of Application analysis integration test
 
Once this PR gets in as a foundation for Konveyor upstream tests, I'm happy discuss other close-future tasks on tests as well as running these tests on PRs/nightly.

jortel reviewed (approved):
LGTM
Thanks for all of the hard work on this!

aufi commented on test/integration/applications-inventory/analysis/windup_basic_test.go:
```diff
@@ -0,0 +1,87 @@
+//
+// Test application analysis
+// "Basic" means that there no other dependencies than the application itself (no need prepare credentials, proxy, etc)
+func TestBasicAnalysis(t *testing.T) {
```
I partially agree, but having this in test/integration directory (instead of test/api), expecting its move to different repo (or at least be just the integration test proving that Hub can triger addon analysis) and having more advanced analysis test in different repos looked OK-ish for me. Also it could be a helpful example for other&QE how to use the test methods from Hub.
Fot that reason I put it to this PR. If this would block this PR from getting in, I can remove it.

aufi commented:
Thanks for your feedback! Merging.
